apps/www/pages/_middleware.tsx

```diff
@@ -0,0 +1,79 @@
+import { NextFetchEvent, NextRequest, NextResponse } from 'next/server'
+import { parseAndVerifyJWT } from '../lib/auth/JWT/JWTHelper'
+import { NativeAppHelpers } from '../lib/withInNativeApp'
+import reportError from '../lib/errors/reportError'
+
+/**
+ * Middleware used to conditionally redirect between the marketing- and front-page
+ * depending on the user authentication status and roles.
+ * @param req
+ * @param ev
+ */
+export async function middleware(req: NextRequest, ev: NextFetchEvent) {
+  const resUrl = req.nextUrl.clone()
+  const userAgent = req.headers.get('user-agent')
+
+  try {
+    // Don't run the middleware unless on home-page
+    if (req.nextUrl.pathname !== '/') {
+      return NextResponse.next()
+    }
+
+    // Parse and verify JWT to decide about redirection
+    const jwtBody = await parseAndVerifyJWT(req)
+    const isMember = jwtBody?.roles.includes('member')
+
+    // Redirect to front-preview ssr to generate article front-preview
+    // used in the yearly overview
+    if (
+      !req.nextUrl.searchParams.has('marketing') &&
+      req.nextUrl.searchParams.has('extractId')
+    ) {
+      // Remap extractId query param to id-slug
+      const extractId = req.nextUrl.searchParams.get('extractId')
+      resUrl.searchParams.delete('extractId')
+
+      resUrl.pathname = `/_ssr/front-preview/${extractId}`
+
+      return NextResponse.rewrite(resUrl)
+    }
+
+    if (isMember) {
+      resUrl.pathname = '/front'
+      return NextResponse.rewrite(resUrl)
+    }
+    // Render marketing-page
+    const isInNativeIOSApp =
+      !!NativeAppHelpers.getIOSVersion(userAgent) &&
+      !!NativeAppHelpers.getNativeAppVersion(userAgent)
+    // Show login instead of marketing page in native app on IOS if version < 2.1.0
+    if (
+      isInNativeIOSApp &&
+      !NativeAppHelpers.isNewerVersion(
+        '2.1.0',
+        NativeAppHelpers.getNativeAppVersion(userAgent),
+      )
+    ) {
+      resUrl.pathname = '/anmelden'
+    } else {
+      resUrl.pathname = '/'
+    }
+    return NextResponse.rewrite(resUrl)
+  } catch (err) {
+    ev.waitUntil(
+      new Promise((resolve) => {
+        fetch(`${req.nextUrl.protocol}//${req.nextUrl.host}/api/reportError`, {
+          headers: {
+            'user-agent': userAgent, // Forward the client user agent
+          },
+          method: 'POST',
+          body: err.stack,
+        })
+          .catch((err) =>
+            console.log('Failed to report error in middleware', err, resUrl),
+          )
+          .finally(() => resolve(null))
+      }),
+    )
```

Comment:
why not just `console.error` here? This always runs on the server and could just do a simple log instead of making an http request that then ends up making a simple log ;-)

Reply:
I shall agree but still be unhappy that we don't use a nice error logging service :P

Reply:
I disagree. While errors in the middleware are most likely to occur due to "server"-side errors it still might be causes due to the request (especially it's headers). In fact, I think we should also log the jwt cookie in the error as-well as the presence of the session cookie. And I still think we should parse the user agent. Non consistent logs are very undesirable. Also the strain on the server is negligible (also it doesn't slow down the middleware since the log to the api-route is made after the middle-ware response has been returned to the user.)

Reply:
Logging (session) cookies is a absolute no go. Also isn't this an bad loop where the request passes through the middleware again? Or are api request excluded from the middleware?

Reply:
Not the session cookie, just the JWT (not usable without the session-cookie anyways). And since it's a post request with HTTPS it would also not be accesible.
Every request goes through the middleware, however this isn't problematic since we have a guard clause that immediatly return if the route is != '/'

Reply:
Not the session cookie, just the JWT (not usable without the session-cookie anyways). And since it's a post request with HTTPS it would also not be accesible.
Every request goes through the middleware, however this isn't problematic since we have a guard clause that immediatly return if the route is != '/'

Reply:
imagine something like:
```
{
  sessionCookiePresent: true,
  jwt: "dasfsdf.adsfafdsfdsaadfs.afsdfdas"
}
```
This would provide us all information we need to see why rewrite might have failed.

Reply:
Honestly I don't get see the point of this whole try catch block – does next docu suggest doing this? Let's discuss later over voice

Reply:
There are countless errors that can occure in the above code, for example during JWT verification/validation.
The main reason why I want to log the errors is that if the middleware pages, both the front-breaks.

Reply:
Moved try-catch arround jwt verification instead of arround the entire middleware